### ISSUE TITLE
Guest touchscreen support

### ISSUE THREAD

**Issue by Jo-Taylor** [QubesOS/qubes-issues#3337]
While Qubes technically does support touchscreen devices it currently does so through XFCE interpreting touchscreen events and generating mouse events from them. This severely limits Qubes utility to, say, host drawing software for stylus supporting devices in GuestVMs, or to utilise the much more developed touchscreen support present in GUIs such as Windows or Gnome.

It seems like it would be fairly simple to add to the Qubes GUI protocol to let it directly grab and pass through touchscreen events rather than relying on Xfce to generate mouse events. This would allow, say, a full screen Windows VM to run OneNote with full palm rejection, or support for automatic detection of text fields and providing a touch keyboard inside of a fullscreen VM, or any number of other touchscreen based use cases on fully touchscreen supporting OSs like Windows or even native Fedora in a HVM, with minimal additional code in the Qubes GUI protocol. This would be particularly useful in the increasingly common 2 in 1 device market, enabling Qubes users to access the same touchscreen features available on other operating systems by simply passing through the appropriate inputs to a VM providing the software intended to receive touchscreen input.

**Comment by a-barinov** [comment]
For the sake of preciseness, the issue seems to have nothing to do with XFCE. I have a touchscreen connected directly to dom0 and it successfully generates touch events (and related mouse events) as evidenced by 'xinput test-xi2 <device>'. However these touch events are not passed to domU by qubesdev device. Looks like support for passing touch events from dom0 to domU should be added for proper guest touchscreen support.

**Comment by v6ak** [comment]
Adding it to Qubes OS seems to require:

* Adding a new message type for touches to [the enum](https://github.com/QubesOS/qubes-gui-common/blob/4fbd87826c558e49b13d36769de59a8bcb0818bb/include/qubes-gui-protocol.h#L70-L97) plus probably some simple struct.
* Adding support for capturing the touches in dom0/GuiVM.
* Adding at least one of those (ideally both):
  * Support in [the GUI agent for Linux](https://github.com/QubesOS/qubes-gui-agent-linux).
  * Support in [HVM stubdom](https://github.com/QubesOS/qubes-gui-agent-xen-hvm-stubdom).

Adding the message type is probably easy once you know what you need (e.g., how are multiple simultaneous touches reported).

Capturing the touch events: It might be easy, but I don't know how to capture it. It seems that touch can be distinguished from mouse-drag (like Qubes Updater do when scrolling the log by finger), but by default, X11 tries to emulate mouse. Mouse events seems to be subscribed [there](https://github.com/QubesOS/qubes-gui-daemon/blob/f6b25d65424345fb4a73922e53053fdf280db9d6/gui-daemon/xside.c#L318) and handled [there](https://github.com/QubesOS/qubes-gui-daemon/blob/f6b25d65424345fb4a73922e53053fdf280db9d6/gui-daemon/xside.c#L1299) through [this call](https://github.com/QubesOS/qubes-gui-daemon/blob/f6b25d65424345fb4a73922e53053fdf280db9d6/gui-daemon/xside.c#L1684). In the documentation of [XSelectInput](https://tronche.com/gui/x/xlib/event-handling/XSelectInput.html) and [event mask](https://tronche.com/gui/x/xlib/window/attributes/event-and-do-not-propagate.html) (or maybe [newer event mask](https://tronche.com/gui/x/xlib/events/processing-overview.html)), I don't see any option for that. I am not sure if Tronche.com is the best source of up-to-date information.

I have seen some note that touch events can be handled through uinput. I am not sure if it can prevent them from being reported as mouse dragging.

**Comment by marmarek** [comment]
Can touch events be represented at the protocol level as extra mouse buttons? We already have coordinates + up/down there.

I have no idea how X11 reports touch events to applications. I see how it works at uinput / X11 input driver level and there are reported there as distinct set of buttons (mostly BTN_TOUCH).

**Comment by v6ak** [comment]
When running xev in dom0, touch is represented as button 1, which is the same as left mouse button.

Also reporting it as mouse buttons IMHO cannot work well with multitouch, unless you have multiple mouse pointers. (Hmm, I vaguely remember that this is probably the way that X11 is designed in.)

**Comment by marmarek** [comment]
There is probably some X11 extension with details about touch. But at the protocol level it still can use mouse buttons, only later translated to proper touch events.

**Comment by v6ak** [comment]
When running xev in dom0, touch is always represented as button 1, which is the same as left mouse button.

Also reporting it as mouse buttons IMHO cannot work well with multitouch, unless you have multiple mouse pointers. (Hmm, I vaguely remember that this is probably the way that X11 is designed in. Ant it looks like I remember it right [[1]](https://wiki.archlinux.org/index.php/Multi-pointer_X) [[2]](https://en.wikipedia.org/wiki/Multi-Pointer_X)). I am having difficulties with writing a real code that reads those values, though. There is xpyb library (or xcffib for Python3, which is mostly the same), but there is not much info how to use it. There is [some not yet published experiment](https://plashless.wordpress.com/2013/11/04/x11-xcb-xpyb-xev-pyxev-for-xi2-touch-events-and-gestures/) by @bootchk with the library.

I am not sure how to move forward at this point.

**Comment by marmarek** [comment]
xi2 reminds me that `xinput test-xi2` exists. Maybe it will help?

**Comment by v6ak** [comment]
You are right. It can distinguish touch events from mouse events and even multiple fingers. I can provide more info, e.g., some fragments of the output.

Not sure what to do the next. The GUID is based on forwarding X11 events from dom0 to domU. As far as I understand, libinput is another library and another API, so using libinput requires us to do it in another way. It might be easy to capture touch events using libinput, but we also need to inject those events in dom0. I am not sure if we can do it in libinput. And even if it is possible, it might be too high-level.

**Comment by marmarek** [comment]
Don't you see those xi2 events in dom0 too?

**Comment by v6ak** [comment]
I see the touch events in dom0 (as expected, as Qubes Updater supports scrolling the log by finger). In domUs, I see the touches as mouse events (also as expected).

**Comment by marmarek** [comment]
Fragments of the output would be useful. Depending on the needs, we could either encode it into existing GUI protocol messages and adjust how we inject them into VM's X server, or create totally new messages.

**Comment by v6ak** [comment]
See: https://gist.github.com/v6ak/ed62c1563272d708f14009a2fd7ed6ac

Observations:

* There are two kinds of touch events – those related to the specific window (`Touch*`) and those related to whole screen (`RawTouch*`). I guess `Touch*` will fit better than `RawTouch*`.
* The window where the touch starts matters: When I start the touch on the window and slide out of it, I still get `Touch*` events. When I start outside the window and slide on the window, I don't get any Touch* event.
* It seems that there is some initial message (DeviceChanged) with some details, including range.
* It looks like field `valuators` are raw values in range mentioned in DeviceChanged and field `root` is scaled to screen resolution (i.e., 1920*1080 in my case).
* Field `detail` seems to be finger identifier. No, I don't have thousands of fingers, the identifier just increases on every touch. It seems to be effectively kind of touch counter.

**Comment by pablob127** [comment]
Adding stylus event support (separate from touchscreen) could be useful, as in many cases they are used for different things (such as stylus for writing and touchscreen for zooming and panning).
I wonder if it would make sense (or if it is even possible) to create new virtual pointer devices exported to the domU rather than coalesce all events into a single one.
